### PR TITLE
pdnsrec 4.0.1

### PR DESCRIPTION
**Pull request by Habbie**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [X] Does your submission pass `brew audit --strict --online <formula>` (after doing `brew install <formula>`)?

---
- update to 4.0.0
- add HEAD support
- remove example config as upstream ships one by default now

### CODE REVIEW

**Comment by DomT4** [comment]
Apparently it wants OpenSSL now:
```
checking for openssl/crypto.h in /usr/local/opt/openssl... no
checking whether compiling and linking against OpenSSL works... no
configure: error: OpenSSL not found
```

If that's assuming Homebrew lives in `/usr/local` we'll probably want to feed it a `--with-openssl` style argument, if possible, here.

File: Formula/pdnsrec.rb [hunk 1]
```diff
@@ -12,40 +12,49 @@ class Pdnsrec < Formula
     sha256 "a9faf9edf0e71de5e9ec9fa70d27811f353a8451b41746208fbdc0d592aa5910" => :mountain_lion
   end
 
+  head do
+    url "https://github.com/powerdns/pdns.git"
+
+    depends_on "automake" => :build
+    depends_on "autoconf" => :build
+    depends_on "libtool"  => :build
```
Comment (DomT4) [inline Formula/pdnsrec.rb]: Please don't bother lining up the `:build` tags. I don't think it makes things that much more readable & some linters/etc groan about it.

File: Formula/pdnsrec.rb [hunk 2]
```diff
@@ -12,40 +12,49 @@ class Pdnsrec < Formula
...
+  depends_on "pkg-config" => :build if build.with? "lua"
+
   def install
-    # Set overrides using environment variables
-    ENV["DESTDIR"] = "#{prefix}"
-    ENV["OPTFLAGS"] = "-O0"
-    ENV.O0
+    args = %W[
+      --prefix=#{prefix}
+      --sysconfdir=#{prefix}/etc/powerdns
```
Comment (DomT4) [inline Formula/pdnsrec.rb]: Should be `#{etc}/powerdns` unless we want to smash people's configuration files every revision/update.

File: Formula/pdnsrec.rb [hunk 3]
```diff
@@ -12,40 +12,49 @@ class Pdnsrec < Formula
...
+      --with-openssl=#{Formula["openssl"].opt_prefix}
+    ]
 
     # Include Lua if requested
     if build.with? "lua"
-      ENV["LUA"] = "1"
-      ENV["LUA_CPPFLAGS_CONFIG"] = "-I#{Formula["lua"].opt_include}"
-      ENV["LUA_LIBS_CONFIG"] = "-llua"
+      args << "--with-lua"
```
Comment (DomT4) [inline Formula/pdnsrec.rb]: Can one-line this now it's not a blob:
```ruby
args << "--with-lua" if build.with? "lua"
```

File: Formula/pdnsrec.rb [hunk 4]
```diff
@@ -12,40 +12,49 @@ class Pdnsrec < Formula
...
-    # Adjust hard coded paths in Makefile
-    inreplace "Makefile.in", "/usr/sbin/", "#{sbin}/"
-    inreplace "Makefile.in", "/usr/bin/", "#{bin}/"
-    inreplace "Makefile.in", "/etc/powerdns/", "#{etc}/powerdns/"
-    inreplace "Makefile.in", "/var/run/", "#{var}/run/"
+    inreplace "pdns_recursor.cc", "SYSCONFDIR", "\"#{etc}/powerdns/\""
+    inreplace "rec_control.cc", "SYSCONFDIR", "\"#{etc}/powerdns/\""
+
+    # Bootstrap?
+    Dir.chdir "pdns/recursordist" if build.head?
```
Comment (DomT4) [inline Formula/pdnsrec.rb]: Don't need to call `Dir` directly here.

File: Formula/pdnsrec.rb [hunk 5]
```diff
@@ -12,40 +12,49 @@ class Pdnsrec < Formula
...
+    system "./bootstrap" if build.head?
```
Comment (DomT4) [inline Formula/pdnsrec.rb]: If you've got two `if build.head?` lines here blob 'em:

```ruby
if build.head?
  blah
  blah
end
```

File: Formula/pdnsrec.rb (hunk 1 above)
Comment (Habbie) [inline Formula/pdnsrec.rb]: Ah - I copied this from another formula. Will fix.

File: Formula/pdnsrec.rb (hunk 4 above)
Comment (Habbie) [inline Formula/pdnsrec.rb]: just `chdir` instead of `Dir.chdir`?

File: Formula/pdnsrec.rb (hunk 2 above)
Comment (Habbie) [inline Formula/pdnsrec.rb]: pdnsrec's `make install` does not write recursor.conf to etcdir, only recursor.conf-dist. Below, at line 47, I actually removed the clobbering that this formula did. If I use #{etc} here, users will have to edit their config in /usr/local/Cellar/... instead of /usr/local/etc/powerdns, as far as I can see.

**Comment by Habbie** [comment]
I'll fix the openssl issue and the nits; awaiting response on the `etc/recursor.conf` stuff.

File: Formula/pdnsrec.rb [hunk 6]
```diff
@@ -12,40 +12,49 @@ class Pdnsrec < Formula
...
+      args << "--with-lua"
     end
 
-    # Adjust hard coded paths in Makefile
-    inreplace "Makefile.in", "/usr/sbin/", "#{sbin}/"
-    inreplace "Makefile.in", "/usr/bin/", "#{bin}/"
-    inreplace "Makefile.in", "/etc/powerdns/", "#{etc}/powerdns/"
-    inreplace "Makefile.in", "/var/run/", "#{var}/run/"
+    inreplace "pdns_recursor.cc", "SYSCONFDIR", "\"#{etc}/powerdns/\""
+    inreplace "rec_control.cc", "SYSCONFDIR", "\"#{etc}/powerdns/\""
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: These replacements shouldn't be necessary. You should be able to set SYSCONFDIR=... in the make/make install.

File: Formula/pdnsrec.rb [hunk 7]
```diff
@@ -12,40 +12,49 @@ class Pdnsrec < Formula
...
-    # Do the install manually
-    bin.install "rec_control"
-    sbin.install "pdns_recursor"
-    man1.install "pdns_recursor.1", "rec_control.1"
-
-    # Generate a default configuration file
-    (prefix/"etc/powerdns").mkpath
-    system "#{sbin}/pdns_recursor --config > #{prefix}/etc/powerdns/recursor.conf"
+  test do
+    output = shell_output("#{sbin}/pdns_recursor --version 2>&1", 0)
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: 0 is already the default

File: Formula/pdnsrec.rb [hunk 8]
```diff
@@ -12,40 +12,49 @@ class Pdnsrec < Formula
...
+
+    # Bootstrap?
+    Dir.chdir "pdns/recursordist" if build.head?
+    system "./bootstrap" if build.head?
 
     # Compile
-    system "./configure"
+    system "./configure", *args
     system "make"
+    system "make", "install"
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Is separating make and make install necessary?

File: Formula/pdnsrec.rb (hunk 6 above)
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Also, you're setting sysconfdir in configure, so why is this needed at all? Is there a bug?

File: Formula/pdnsrec.rb [hunk 9]
```diff
@@ -12,40 +12,49 @@ class Pdnsrec < Formula
...
+
   def install
-    # Set overrides using environment variables
-    ENV["DESTDIR"] = "#{prefix}"
-    ENV["OPTFLAGS"] = "-O0"
-    ENV.O0
+    args = %W[
+      --prefix=#{prefix}
+      --sysconfdir=#{prefix}/etc/powerdns
+      --with-openssl=#{Formula["openssl"].opt_prefix}
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Configure can't find this on its own?

File: Formula/pdnsrec.rb (hunk 4 above)
Comment (Habbie) [inline Formula/pdnsrec.rb]: just `cd` according to @ilovezfs

**Comment by Habbie** [comment]
I'll do another iteration to re-figure out what the story with SYSCONFDIR etc. was so you get a clear picture.

File: Formula/pdnsrec.rb (hunk 5 above)
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: I think we can avoid that. See suggestion above.

File: Formula/pdnsrec.rb (hunk 4 above)
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: I'm not thrilled with two aspects of this
a) the only-if-head directory change
b) the directory change is permanent not in a block

I'd be tempted to do this instead:
```
  def install
    (buildpath/"pdns/recursordist").install buildpath.children if build.stable?

    ...

    cd "pdns/recursordist" do
      inreplace %w[pdns_recursor.cc rec_control.cc],
        "SYSCONFDIR",
        "\"#{etc}/powerdns/\""                        
      system "./bootstrap" if build.head?
      system "./configure", *args
      system "make"
      system "make", "install"
    end
```

File: Formula/pdnsrec.rb (hunk 6 above)
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Note that for HEAD these would need to be after the directory change

File: Formula/pdnsrec.rb (hunk 4 above)
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: And here's an even better idea: we could just remove the heads spec for this thing.

Comment (DomT4) [inline Formula/pdnsrec.rb]: ILZ is correct on our `cd` preference, but I didn't see a way to use `cd` here in our preferred `do` format without making this formula silly complex to visually parse.

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: I think we should just kill the head spec. It's a subfolder within the pdns repository, which is what's causing the madness, and we already have a separate pdns formula that has its own head spec.

Comment (Habbie) [inline Formula/pdnsrec.rb]: The head spec in the pdns formula installs a different product, to be clear.

Comment (DomT4) [inline Formula/pdnsrec.rb]: Yeah, I don't necessarily disagree with you.

It seems like there's a lot we'd need to do to accommodate `head` here. We could look at doing so, but perhaps it's worth waiting to see if we get a sudden influx of demand for `head` on this formula.

Comment (Habbie) [inline Formula/pdnsrec.rb]: I'll take out the HEAD stuff for now, it will simplify things. Prior to this 4.0.0 release there was some demand in fact, but now that 4.0.0 is finally out (after 16 months) demand should be limited.

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: >sudden influx of demand for head on this formula.

LOL LOL LOL

Comment (DomT4) [inline Formula/pdnsrec.rb]: > LOL LOL LOL

I added "on this formula" in the hope people wouldn't burst out laughing at the phrasing 😅.

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: @Habbie are you saying `make install` doesn't install recursor at all if you're doing `brew install --HEAD pdns`?

Comment (Habbie) [inline Formula/pdnsrec.rb]: Yes, I am. pdns_server, pdns-recursor and dnsdist are distinct products that live in one git tree because they share a lot of code. The non-HEAD versions are distributed as tarballs that are each individually complete (and thus overlap a bit), but when building from git, getting dnsdist or the recursor requires a chdir first.

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: I see.

@DomT4 The other option would be to require Autotools for stable, and build from the source archive tarball, which would bring them both into alignment without further fuss.

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Anyway, I can live with the random directory change for only head but it's pretty yucky.

Comment (DomT4) [inline Formula/pdnsrec.rb]: > The other option would be to require Autotools for stable

Heavily prefer not to do this as much as possible.

Comment (Habbie) [inline Formula/pdnsrec.rb]: I'll just take HEAD out, it's fine. Expect a version with all your comments honoured within a day or two.

**Comment by Habbie** [comment]
So, if I do `--sysconfdir=#{etc}/powerdns`, recursor correctly compiles with this default `# config-dir=/usr/local/etc/powerdns` but it installs `recursor.conf-dist` directly into `/usr/local/etc/powerdns/` which I think is a violation of homebrew policy?

If I do `--sysconfdir=#{prefix}/etc/powerdns`, I get `/usr/local/etc/powerdns//recursor.conf-dist -> ../../Cellar/pdnsrec/4.0.0/etc/powerdns/recursor.conf-dist` which is good, but the recursor gets this configuration default `config-dir=/usr/local/Cellar/pdnsrec/4.0.0/etc/powerdns` which is bad, as it requires the user to put their config there - or we need a `recursor.conf` symlink in `/usr/local/etc/powerdns` which feels like the kind of thing an editor might accidentally destroy. It would also not survive upgrades.

So, I feel that my only choice here is to do the latter (`--sysconfdir=#{prefix}/etc/powerdns`) plus patch those two binaries so they get the right `config-dir`. Thoughts?

**Comment by Habbie** [comment]
For your review, I have pushed a new version. HEAD is gone, openssl and boost are detected properly. The config path issue is not fully fixed (see previous comment for options). Some commented lines are still in there as this is unfinished.

File: Formula/pdnsrec.rb (hunk 2 above)
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Setting it to #{etc}/powerdns will not clobber existing files. Then you should handle the upgrade/replace/merge/whatever logic in a `post_install` method.

**Comment by ilovezfs** [comment]
>recursor.conf-dist directly into /usr/local/etc/powerdns/ which I think is a violation of homebrew policy?

It's the reverse. The `etc` and `var` pathnames are exceptional in that they deliberately point into the `HOMEBREW_PREFIX` instead of into the formula `prefix`. Then in `post_install` (no sandbox and post bottle pouring) you make sure that the relevant directories are created in `var` (e.g., `(var/"run/foo").mkpath`) if needed, and that the config files are upgraded/merged/replaced/renamed/whatever in `etc` or `etc/"foo"` if the files were installed with the incorrect name or if they weren't installed at all because brew refused to overwrite a preexisting file.

**Comment by Habbie** [comment]
@ilovezfs so `--sysconfdir=#{etc}/powerdns` and no patching would be fine?

File: Formula/pdnsrec.rb (hunk 2 above)
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: ping.

File: Formula/pdnsrec.rb [hunk 10]
```diff
@@ -14,38 +14,34 @@ class Pdnsrec < Formula
...
-    inreplace "Makefile.in", "/usr/sbin/", "#{sbin}/"
-    inreplace "Makefile.in", "/usr/bin/", "#{bin}/"
-    inreplace "Makefile.in", "/etc/powerdns/", "#{etc}/powerdns/"
-    inreplace "Makefile.in", "/var/run/", "#{var}/run/"
+    args << "--with-lua" if build.with? "lua"
 
-    # Compile
-    system "./configure"
-    system "make"
+    # inreplace "pdns_recursor.cc", "SYSCONFDIR", "\"#{etc}/powerdns/\""
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: delete

File: Formula/pdnsrec.rb [hunk 11]
```diff
@@ -14,38 +14,34 @@ class Pdnsrec < Formula
...
+    # inreplace "rec_control.cc", "SYSCONFDIR", "\"#{etc}/powerdns/\""
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: delete

File: Formula/pdnsrec.rb [hunk 12]
```diff
@@ -14,38 +14,34 @@ class Pdnsrec < Formula
...
 
-    # Do the install manually
-    bin.install "rec_control"
-    sbin.install "pdns_recursor"
-    man1.install "pdns_recursor.1", "rec_control.1"
+    # Compile
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: delete

File: Formula/pdnsrec.rb [hunk 13]
```diff
@@ -14,38 +14,34 @@ class Pdnsrec < Formula
...
+    system "./configure", *args
+    # system "make"
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: delete

File: Formula/pdnsrec.rb [hunk 14]
```diff
@@ -14,38 +14,34 @@ class Pdnsrec < Formula
 
   depends_on :macos => :lion
   depends_on "boost"
+  depends_on "openssl"
   depends_on "lua" => :optional
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: I wonder if we should just require this and kill the option.

Comment (Habbie) [inline Formula/pdnsrec.rb]: Works for me. Doing that.

File: Formula/pdnsrec.rb (hunk 2 above)
Comment (Habbie) [inline Formula/pdnsrec.rb]: sorry, missed this comment.

**Comment by Habbie** [comment]
Updated. Lua is no longer optional. Removed all the comments. Went back to `--sysconfdir=#{etc}/powerdns`. I hope this is done now!

**Comment by ilovezfs** [comment]
@Habbie I don't think your commit made it to the mothership.

**Comment by Habbie** [comment]
It kind of did but not with the changes. Fixed now. Apologies.

**Comment by MikeMcQuaid** [comment]
Fails to build on 10.9: http://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/4947/version=mavericks/testReport/junit/brew-test-bot/mavericks/install_pdnsrec/

**Comment by Habbie** [comment]
Bumped version requirement to 10.10.

**Comment by ilovezfs** [comment]
@Habbie I'd suggest trying this first:
```
diff --git a/Formula/pdnsrec.rb b/Formula/pdnsrec.rb
index 0b076a4..7b7183b 100644
--- a/Formula/pdnsrec.rb
+++ b/Formula/pdnsrec.rb
@@ -12,14 +12,17 @@ class Pdnsrec < Formula
     sha256 "a9faf9edf0e71de5e9ec9fa70d27811f353a8451b41746208fbdc0d592aa5910" => :mountain_lion
   end
 
-  depends_on :macos => :yosemite
+  depends_on "pkg-config" => :build
   depends_on "boost"
   depends_on "openssl"
   depends_on "lua"
+  depends_on "gcc" => :linked if MacOS.version <= :mavericks
 
-  depends_on "pkg-config" => :build
+  needs :cxx11
 
   def install
+    ENV.cxx11
+
     args = %W[
       --prefix=#{prefix}
       --sysconfdir=#{etc}/powerdns
```

**Comment by Habbie** [comment]
Did that, mavericks still unhappy. Looks like it's using clang though - and you tried to make it use gcc it looks like.

**Comment by Habbie** [comment]
Added `fails_with :clang` block to force gcc usage on Mavericks. Bumped to 4.0.1. Please review again!

File: Formula/pdnsrec.rb [hunk 15]
```diff
@@ -12,40 +12,37 @@ class Pdnsrec < Formula
...
+  depends_on "pkg-config" => :build
   depends_on "boost"
-  depends_on "lua" => :optional
+  depends_on "openssl"
+  depends_on "lua"
+  depends_on "gcc" => :linked if MacOS.version <= :mavericks
+
+  needs :cxx11
+
+  fails_with :clang do
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: is this needed anywhere but Mavericks?

File: Formula/pdnsrec.rb [hunk 16]
```diff
@@ -12,40 +12,37 @@ class Pdnsrec < Formula
     sha256 "a9faf9edf0e71de5e9ec9fa70d27811f353a8451b41746208fbdc0d592aa5910" => :mountain_lion
   end
 
-  depends_on :macos => :lion
+  depends_on "pkg-config" => :build
   depends_on "boost"
-  depends_on "lua" => :optional
+  depends_on "openssl"
+  depends_on "lua"
+  depends_on "gcc" => :linked if MacOS.version <= :mavericks
```
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Let's kill this entire line.

File: Formula/pdnsrec.rb (hunk 15 above)
Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Ah, I see that's the point of the 600.

File: Formula/pdnsrec.rb (hunk 16 above)
Comment (Habbie) [inline Formula/pdnsrec.rb]: If I kill that line, gcc does not get installed and pdnsrec does not build on mavericks

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: That's handled by the fails_with.

Comment (Habbie) [inline Formula/pdnsrec.rb]: Hmm, pretty sure it wasn't for me. Will give that another shot.

Comment (Habbie) [inline Formula/pdnsrec.rb]: If I do
```diff
-  depends_on "gcc" => :linked if MacOS.version <= :mavericks
+  #depends_on "gcc" => :linked if MacOS.version <= :mavericks
```
I get
```
Error: pdnsrec cannot be built with any available compilers.
Install GNU's GCC
  brew install gcc
```

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Run test-bot instead of brew install :)

Comment (Habbie) [inline Formula/pdnsrec.rb]: (1) how do I do that? (2) won't that still leave Mavericks users with a confusing error (if we ignore the presence of bottles)?

Comment (ilovezfs) [inline Formula/pdnsrec.rb]:
```
brew pull 2895
export HOMEBREW_DEVELOPER=1
export HOMEBREW_SANDBOX=1
brew test-bot pdnsrec
```
No, it just gives them opportunity to decide whether they want to install gcc.

Comment (DomT4) [inline Formula/pdnsrec.rb]: > Let's kill this entire line.

It actually links to `gcc` in the bottle, so I'm not sure ditching the dep was wise. It may cause some of the `linkage` issues we'd discussed before:

```
@@HOMEBREW_PREFIX@@/opt/gcc/lib/gcc/6/libstdc++.6.dylib
```

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: I know it does, but I do not consider that a sufficient reason for a hard dependency on gcc in the formula for Mavericks, unless it could be scoped to the bottle, which it can't be. Fails_with is supposed to provide flexibility not lock in a specific compiler dependency.

Comment (DomT4) [inline Formula/pdnsrec.rb]: > Fails_with is supposed to provide flexibility not lock in a specific compiler dependency.

Sure, ideally. It doesn't though, really, due to the way we handle this for bottles. It won't be any issue once we stop shipping Mavericks bottles in the near future anyway, so meh.

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: >Sure, ideally.

It's test-bot that's rescuing `CompilerSelectionError` and then making a choice to install gcc as a remedy. If that behavior causes us to add `depends_on "gcc"` in the formula, it's the tail wagging the dog.

Comment (DomT4) [inline Formula/pdnsrec.rb]: Ship it as-is. We'll see how we go.

Comment (ilovezfs) [inline Formula/pdnsrec.rb]: Sounds good. :dog:

**Comment by ilovezfs** [comment]
@Habbie Nice work on this. I like how it turned out! Thanks for another contribution to Homebrew and your patience with the tag teaming.

**Comment by Habbie** [comment]
Thanks for the work, to both of you!
